solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,376 @@
+        // ITokenBridge.completeTransferWithPayload completes a contract-controlled
+        // transfer of an ERC20 token. Calling this function is not enough to
+        // ensure L2WormholeGateway received Wormhole tBTC representation.
+        // Instead of going too deep into the ITokenBridge implementation,
+        // asserting who is the receiver of the token, and which token it is,
+        // we check the balance before the ITokenBridge call and the balance
+        // after ITokenBridge call. This way, we are sure this contract received
+        // Wormhole tBTC token in the given amount. This is transparent to
+        // all potential upgrades of ITokenBridge implementation and no other
+        // validations are needed.
+        uint256 balanceBefore = bridgeToken.balanceOf(address(this));
+        bytes memory encoded = bridge.completeTransferWithPayload(encodedVm);
+        uint256 balanceAfter = bridgeToken.balanceOf(address(this));
+
+        uint256 amount = balanceAfter - balanceBefore;
+        // Protect against the custody of irrelevant tokens.
+        require(amount > 0, "No tBTC transferred");
+
+        address receiver = fromWormholeAddress(
+            bytes32(bridge.parseTransferWithPayload(encoded).payload)
+        );
+        require(receiver != address(0), "0x0 receiver not allowed");
+
+        // We send wormhole tBTC OR mint canonical tBTC. We do not want to send
+        // dust. Sending wormhole tBTC is an exceptional situation and we want
+        // to keep it simple.
+        if (mintedAmount + amount > mintingLimit) {
+            bridgeToken.safeTransfer(receiver, amount);
+        } else {
+            // The function is non-reentrant given bridge.completeTransferWithPayload
+            // call that does not allow to use the same VAA again.
+            // slither-disable-next-line reentrancy-benign
+            mintedAmount += amount;
+            tbtc.mint(receiver, amount);
+        }
+
+        // The function is non-reentrant given bridge.completeTransferWithPayload
+        // call that does not allow to use the same VAA again.
+        // slither-disable-next-line reentrancy-events
+        emit WormholeTbtcReceived(receiver, amount);
+    }
+
+    /// @notice Allows to deposit Wormhole tBTC token in exchange for canonical
+    ///         tBTC. Useful in a situation when user received wormhole tBTC
+    ///         instead of canonical tBTC. One example of such situation is
+    ///         when the minting limit was exceeded but the user minted anyway.
+    /// @dev Requirements:
+    ///      - The sender must have at least `amount` of the Wormhole tBTC and
+    ///        it has to be approved for L2WormholeGateway.
+    /// @param amount The amount of Wormhole tBTC to deposit.
+    function depositBridgeToken(uint256 amount) external {
+        bridgeToken.transferFrom(msg.sender, address(this), amount);
+        tbtc.mint(msg.sender, amount);
```

Comment:
Does the `mintedAmount` state variable need to be incremented here?

Reply:
Yep, good catch. Added in da36bed40c381e94527e3265880988b7fad5c21b.